DigitalLearningSolutions.Data/Services/PasswordResetService.cs

```diff
@@ -0,0 +1,121 @@
+            this.connection = connection;
+            this.logger = logger;
+            this.configService = configService;
+            this.emailService = emailService;
+        }
+
+        public void SendResetPasswordEmail(string emailAddress)
+        {
+            User user =
+                this.userService.GetUserByEmailAddress(emailAddress) ??
+                throw new EmailAddressNotFoundException("No user account could be found with the specified email address");
+            string resetPasswordHash = GenerateResetPasswordHash(user);
+            Email resetPasswordEmail = GeneratePasswordResetEmail(emailAddress, resetPasswordHash, user.FirstName);
+            this.emailService.SendEmail(resetPasswordEmail);
+        }
+
+        private string GenerateResetPasswordHash(User user)
+        {
+            string hash = Guid.NewGuid().ToString();
+            string tableName = user.GetType() == typeof(DelegateUser) ? DelegatesTableName : AdminTableName;
+            string idColumnName = user.GetType() == typeof(DelegateUser) ? DelegatesIdColumnName : AdminIdColumnName;
+
+            int numberOfAffectedRows = connection.Execute(
+                $@"BEGIN TRY
+                        DECLARE @ResetPasswordID INT
+                        BEGIN TRANSACTION
+                            INSERT INTO dbo.ResetPassword
+                                ([ResetPasswordHash]
+                                ,[PasswordResetDateTime])
+                            VALUES(@ResetPasswordHash, GETDATE())
+
+                            SET @ResetPasswordID = SCOPE_IDENTITY()
+
+                            UPDATE {tableName}
+                            SET ResetPasswordID = @ResetPasswordID
+                            WHERE {idColumnName} = @UserID
+                        COMMIT TRANSACTION
+                    END TRY
+                    BEGIN CATCH
+                        ROLLBACK TRANSACTION
+                    END CATCH
+                    ", new { ResetPasswordHash = hash, UserID = user.Id });
+            if (numberOfAffectedRows < 2)
+            {
+                string message = $"Not saving reset password hash as db insert/update failed for User ID: {user.Id} from table {tableName}";
+                logger.LogWarning(message);
+                throw new ResetPasswordInsertException(message);
+            }
+
+            return hash;
+        }
+
+        private Email GeneratePasswordResetEmail(string emailAddress, string resetHash, string firstName)
+        {
+            string baseUrl =
+                configService.GetConfigValue(ConfigService.BaseUrl) ??
+                throw new ConfigValueMissingException("Encountered an error while trying to generate a Reset Password email: " +
+                                                      "The value of baseURL is null ");
+            UriBuilder resetPasswordUrl = new UriBuilder(baseUrl);
+            resetPasswordUrl.Path += "ResetPassword";
```

Comment:
Just to check - is the baseUrl for old or refactored site? (We will need the refactored site)

Reply:
This is drawn from the BaseURL (ConfigID=21) located in the Config table. In the backed-up version of the database, it is recorded as "https://www.dls.nhs.uk/". Do you know if we still plan on using that config table for the baseURL, or should there be a different source/value for testing?

Reply:
Ah - I think you can get the url from the config for the refactored site:
Configuration["AppRootPath"]

It looks like the old site is also stored in config:
Configuration["CurrentSystemBaseUrl"]